### ISSUE TITLE
Store locations and original formatting in the AST

### ISSUE THREAD

**Issue by scripthunter7** [AdguardTeam/AGLint#18]
Currently, AST only stores rules data, but not locations and formatting.

## Linter issues
This is a problem because if the following adblock rule is given in the linter:
```adblock
example.com/ads.js$asd
```
then we can only "underline" the entire rule with an error message that says "asd" is an invalid modifier. But it would be a more favorable behavior if, instead of the entire rule, we underlined only the problematic part.

Or if we have the following rule:
```adblock
example.com##.something:contains(aaa
```
then we get an error message from the parser that the closing parenthesis is missing, but the linter takes this as a fatal error, since it happens at the parser level.

However, the parser could be so "tolerant" that it accepts the wrong selector, and we will validate it at a later level. In this way, it would also be possible to underline only the faulty selector in red in VSCode.

In practice, this means that we store the raw selector, which can be validated / modified at any time with CSSTree.

## Generation issues (lossless generation if possible)

Currently, if we regenerate a rule from the parsed AST, the formatting is lost. This could be solved by storing spaces (see PostCSS AST - see https://postcss.org/api/#atrule-raws).

## Summary

To solve these problems, the parser needs to store the locations and the raw parts. So i modified the parser a little bit.

Let this be our example code:
```js
import { DomainListParser } from "aglint";

console.log(DomainListParser.parse("  example.com,   ~example.org"));
```

Until now, when we parsed a domain list, we got the following result:
```json
{
  "type": "DomainList",
  "separator": ",",
  "domains": [
    { "domain": "example.com", "exception": false },
    { "domain": "example.org", "exception": true }  
  ]
}
```

and we were able to generate the following result from it:
```
example.com,~example.org
```

_(formatting are lost)_

However, the reworked parser gives much more detailed results (TEST result):
```json
{
  "type": "DomainList",
  "raw": {
    "before": "  ",
    "after": "",
    "text": "example.com,   ~example.org"
  },
  "location": {
    "start": 0,
    "end": 29
  },
  "separator": ",",
  "domains": [
    {
      "type": "Domain",
      "raw": {
        "before": "",
        "after": "",
        "text": "example.com"
      },
      "location": {
        "start": 2,
        "end": 13
      },
      "value": {
        "type": "Value",
        "raw": {
          "before": "",
          "after": "",
          "text": "example.com"
        },
        "location": {
          "start": 2,
          "end": 13
        },
        "value": "example.com"
      },
      "exception": false
    },
    {
      "type": "Domain",
      "raw": {
        "before": "   ",
        "after": "",
        "text": "~example.org"
      },
      "location": {
        "start": 17,
        "end": 32
      },
      "value": {
        "type": "Value",
        "raw": {
          "before": "",
          "after": "",
          "text": "example.org"
        },
        "location": {
          "start": 17,
          "end": 32
        },
        "value": "example.org"
      },
      "exception": true
    }
  ]
}
```

_(Since an adblock rule can have a maximum of one line, it is enough to store only the columns)_

You can disable storing these additional informations by passing `{ locations: false, raws: false }` to the parser. In such a case, you will receive the following AST:
```json
{
  "type": "DomainList",
  "separator": ",",
  "domains": [
    {
      "type": "Domain",
      "value": {
        "type": "Value",
        "value": "example.com"
      },
      "exception": false
    },
    {
      "type": "Domain",
      "value": {
        "type": "Value",
        "value": "example.org"
      },
      "exception": true
    }
  ]
}
```

And from this we can generate a result equivalent to the original. Raw/location parts are optional. Based on the locations, we can tell VSCode exactly which part of the code to mark as faulty. In addition, the parser is much more tolerant this way, which is favorable for the linter.

@ameshkov
